File: 백준/Silver/14501. 퇴사/퇴사.py
import sys

N = int(input())
d = [0]*(N+2)
Time = [0]
Pay = [0]

for a in range(N):
    puts = sys.stdin.readline().split()
    Time.append(int(puts[0]))
    Pay.append(int(puts[1]))



for a in range(1,N+1):
    for i in range(a+Time[a] , N+2):
        if d[i] < d[a]+Pay[a]:
            d[i] = d[a] + Pay[a]
       



print(d[-1])

File: 백준/Silver/14501. 퇴사/퇴사.java
import java.util.*;
import java.io.*;

public class Main{
  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    int N = Integer.parseInt(br.readLine());
    int[] time = new int[N+1];
    int[] pay = new int[N+1];
    int[] dp = new int[N+1];

    for(int i = 0; i < N; i++)
    {
      StringTokenizer st = new StringTokenizer(br.readLine());
      time[i] = Integer.parseInt(st.nextToken());
      pay[i] = Integer.parseInt(st.nextToken());
    }

    for(int i = 0; i < N+1; i++)
    {
      for(int j = i+time[i]; j < N+1; j++)
      {
        if(dp[j] < dp[i] + pay[i]) dp[j] = dp[i] + pay[i];
      }
    }

    bw.write(dp[N]+"");
    bw.flush();

  }
}